pkg/controllers/selection/preferences.go

```diff
@@ -49,26 +49,34 @@ func NewPreferences() *Preferences {
 // scheduled to that zone. Preferences are removed iteratively until only hard
 // constraints remain. Pods relaxation is reset (forgotten) after 5 minutes.
 func (p *Preferences) Relax(ctx context.Context, pod *v1.Pod) {
-	affinity, ok := p.cache.Get(string(pod.UID))
+	spec, ok := p.cache.Get(string(pod.UID))
 	// Add to cache if we've never seen it before
 	if !ok {
-		p.cache.SetDefault(string(pod.UID), pod.Spec.Affinity)
+		// Limit cached PodSpec to only required data
+		cachedSpec := v1.PodSpec{
+			Affinity:    pod.Spec.Affinity,
+			Tolerations: pod.Spec.Tolerations,
+		}
+		p.cache.SetDefault(string(pod.UID), cachedSpec)
```

Comment:
There are lots of pointers in this nested struct. How does equality checking work?

Reply:
What if two pods have the same spec? I'm not sure we want to relax both of them.

Reply:
> There are lots of pointers in this nested struct. How does equality checking work?

I agree using an empty PodSpec is probably not ideal.  Lots of empty fields there.  I'll define a custom type instead.

> What if two pods have the same spec? I'm not sure we want to relax both of them. This could potentially cause over-relaxation if we see 10 identical pods all at once.

I'm not totally sure I follow.  The pod.UID should always be unique, but maybe that's not what you're getting at.